Upgrade to TypeScript 2.1

alexeagle opened the pull request:
**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)



**What is the new behavior?**



**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

alexeagle commented:
Note, this is currently blocked on a TS 2.1-compatible tsickle.
/cc @evmar

chuckjaz commented:
The changes to TypeScript broke `ts_plugin.ts`. Feel free to remove it and the corresponding tests as it is currently not used.

chuckjaz reviewed (dismissed):
Fix or remove ts_plugin.ts

imcotton commented on tools/@angular/tsc-wrapped/src/collector.ts:
```diff
@@ -12,6 +12,14 @@ import {Evaluator, errorSymbol, isPrimitive} from './evaluator';
 
+// ModifierFlags introduced in TS 2.1
+// TypeScript prior to 2.1 uses the NodeFlags to store modifiers
+const isStatic = !!ts.ModifierFlags ?
```
If set minimum dependency to TS v2.1, then these checking does not necessary.

alexeagle replied:
But Angular 2.x needs to support TypeScript 1.8 because of semver.

imcotton replied:
Oh I see, it's `devDependencies` in `package.json`, so there could be a test scenario to simulates TS v1.8 as well?

alexeagle commented:
We do test the .d.ts files we publish that they work with TS 1.8
https://github.com/angular/angular/blob/master/tools/typings-test/test.sh#L22
But I think we are missing a test that our JS code executes correctly with
TS 1.8 as a runtime dep

On Thu, Dec 8, 2016 at 10:41 AM Cotton Hou <notifications@github.com> wrote:

*@imcotton* commented on this pull request.
------------------------------

In tools/@angular/tsc-wrapped/src/collector.ts
<https://github.com/angular/angular/pull/13294>:

> @@ -12,6 +12,14 @@ import {Evaluator, errorSymbol, isPrimitive} from './evaluator';
 import {ClassMetadata, ConstructorMetadata, FunctionMetadata,
MemberMetadata, MetadataEntry, MetadataError, MetadataMap,
MetadataObject, MetadataSymbolicBinaryExpression,
MetadataSymbolicCallExpression, MetadataSymbolicExpression,
MetadataSymbolicIfExpression, MetadataSymbolicIndexExpression,
MetadataSymbolicPrefixExpression, MetadataSymbolicReferenceExpression,
MetadataSymbolicSelectExpression, MetadataSymbolicSpreadExpression,
MetadataValue, MethodMetadata, ModuleExportMetadata, ModuleMetadata,
VERSION, isClassMetadata, isConstructorMetadata, isFunctionMetadata,
isMetadataError, isMetadataGlobalReferenceExpression,
isMetadataSymbolicExpression, isMetadataSymbolicReferenceExpression,
isMetadataSymbolicSelectExpression, isMethodMetadata} from './schema';
 import {Symbols} from './symbols';

+// ModifierFlags introduced in TS 2.1
+// TypeScript prior to 2.1 uses the NodeFlags to store modifiers
+const isStatic = !!ts.ModifierFlags ?

Oh I see, it's devDependencies in package.json, so there could be a test
scenario to simulates TS v1.8 as well?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/angular/angular/pull/13294>, or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAC5I5YUYomlSN8NDq-PmvhpVfdNKEWVks5rGE9XgaJpZM4LHQfe>
.

chadbr commented:
Everyone go on vacation? 😄

alexeagle commented:
This is blocked on getting angular/tsickle updated to TS 2.1 which is
somewhat harder.

On Thu, Dec 15, 2016 at 7:35 AM chadbr <notifications@github.com> wrote:

> Everyone go on vacation? 😄
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/13294#issuecomment-267357654>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC5I2IPSQPQzPzFSozrn1MX7j1cwOFRks5rIV4zgaJpZM4LHQfe>
> .
>

chadbr commented:
thanks for the update - angular/tsickle#295

Martin-Wegner commented:
Any news?

alexeagle commented:
We are making progress on a TS 2.1 branch of tsickle right now.

mmc41 commented:
@alexeagle According to the angular [changelog](https://github.com/angular/angular/blob/master/CHANGELOG.md) typescript 2.1 support came in 2.4.2 or was that only partial support ?

duydao commented:
@mmc41 Thats for builds without the AOT compiler

jinder commented:
Looks like this landed in tsickle last night:  https://github.com/angular/tsickle/issues/295

fxck commented:
Is this going to be a part of 2.x.x, or it will have some big breaking changes a be part of 4.x.x only?

maxime1992 commented:
@fxck as far I as understood, typescript 2.1.x will only be available in angular v4.x

mmc41 commented:
Would very much appreciate typescript 2.1.x support for 2.x.x

db6edr commented:
Doesn't seem to be possible, because it would be a breaking change. With
semver in mind, this triggers a change of the major version - and that will
be 4.x.

2017-01-24 10:47 GMT+01:00 Morten Christensen <notifications@github.com>:

> Would very much appreciate typescript 2.1.x support for 2.x.x
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/13294#issuecomment-274756317>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAQIjsKEq3pxLSVRwy3yNq0vS9DGE_jPks5rVcjDgaJpZM4LHQfe>
> .
>



-- 
Dirk Räder
Morissestr. 7
26180 Rastede
T: +49-4402-5157715
M: +49-176-60835036
E: dirk@db6edr.de

fxck commented:
Would it be a breaking change though?

db6edr commented:
It seems so: Angular uses tsickle as part of the AoT process. In
https://github.com/angular/tsickle/issues/295, they state that it would not
be possible to support both pre-2.1-TypeScript and 2.1-TypeScript at the
same time.

So the minimum version of a dependency must be changed, which in turn means
a breaking change.

2017-01-24 12:20 GMT+01:00 Aleš <notifications@github.com>:

> Would it be a breaking change though?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/13294#issuecomment-274777177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAQIjnl8MIkXVqw0Q-0TyYO7z4JLa83Mks5rVd6YgaJpZM4LHQfe>
> .
>



-- 
Dirk Räder
Morissestr. 7
26180 Rastede
T: +49-4402-5157715
M: +49-176-60835036
E: dirk@db6edr.de

pglazkov commented:
Why AOT with TypeScript 2.1 works in https://github.com/AngularClass/angular2-webpack-starter? Will it fail only in certain scenarios?

alexeagle commented:
This PR will only target 4.x.
Indeed, Angular 2.x must support TS 1.8 due to semver, but it's possible to use reflection to determine which version is present so we can work with any of 1.8-2.1.
We have a plan to backport the needed bits to have a release of 2.x that can do this, with some manual steps required (change the version of tsickle). I'm not confident we'll have time, but will try.

alexeagle committed:
build: upgrade to TypeScript 2.1

alexeagle committed:
build: remove post-processing of .d.ts files

Note that the duplication in tsconfig.json files will be fixed in a followup CL now that we have tsconfig inheritance

BREAKING CHANGE: Angular 4 will support only TypeScript 2.1, so we no longer provide backwards compatibility to TS 1.8.

alexeagle committed:
fix: build and test fixes for TS 2.1

chuckjaz commented on tools/@angular/tsc-wrapped/src/collector.ts:
```diff
@@ -123,7 +123,13 @@ export class MetadataCollector {
         classDeclaration.heritageClauses.forEach((hc) => {
           if (hc.token === ts.SyntaxKind.ExtendsKeyword && hc.types) {
-            hc.types.forEach(type => result.extends = referenceFrom(type.expression));
+            hc.types.forEach(type => {
```
Wouldn't it make more sense to ignore this in the static reflector instead of here?

alexeagle replied:
yes, it was just harder for me to track down. I'll try that.
Do you think it should be a specific exception for extending Error, or can we always ignore errors from 'extends' metadata?

shlomiassaf commented:
@alexeagle I didn't understand the why AOT compilation works in [AngularClass/angular2-webpack-starter](https://github.com/AngularClass/angular2-webpack-starter)

It also works on my projects and I have used mapped types to ensure its 2.1.  
Actually, I have a project with **typescript nightly 2.2** that I just compiled AOT.

I don't really understand the limitation.

Perhaps it's related to the fact what we use `ngc-webpack` and no the `AOTPlugin`?

chuckjaz commented on modules/@angular/compiler/src/aot/static_reflector.ts:
```diff
@@ -106,6 +111,7 @@ export class StaticReflector implements ReflectorReader {
       const classMetadata = this.getTypeMetadata(type);
       propMetadata = {};
       if (classMetadata['extends']) {
+        console.log("simplify for propMetadata");
```
Left-over debugging statements?

alexbyk commented:
I don't understand. I just created a new project with angular-cli, changed "target" to "es6" and TS 2 works including async/await. Should I be worried and switched back to es5+ts1.8?

alexeagle commented:
Sorry, I'm not up to date on what issues currently exist with Angular 2 + TS 2.1 in different tooling. `@angular/tsc-wrapped` does work, so long as you install `tsickle@0.21`. If it works for you already, then great.

fxck commented:
Well it works fine for me as well, the only problem I noticed is that vscode language service plugin stopped working..

chuckjaz commented:
@fxck Can you report that as an issue against the vscode-ng-language-service project?

fxck commented:
Already did https://github.com/angular/vscode-ng-language-service/issues/41

halfzebra commented:
@pglazkov It fails silently during AoT compilation if you use any of the 2.1 features.

pglazkov commented:
@halfzebra It does work for me, I tried it with the "Object Spread" feature. Also see the comment from @shlomiassaf, he also mentions it it works with the mapped types feature.

kylecordes commented:
I wonder if there is any appetite for an enhancement to A2.x that would make whatever problems occur, obvious up front instead of silently later? It seems like a low urgency (A4 should ship soon anyway according to the development cadence announcement a while back) but it might help out some people who otherwise get tripped up by a silent failure in certain (somewhat unspecified) cases under AOT?

alexeagle commented:
we have upgraded the version of TS inside google so this is unblocked now

beeman commented:
@alexeagle does that mean it will be supported with Angular 2.4.x?

alexeagle commented:
Please see earlier comments about TS 2.1 with NG 2.4, it is reported to
work already.

On Fri, Feb 3, 2017 at 4:04 PM Bram Borggreve <notifications@github.com>
wrote:

> @alexeagle <https://github.com/alexeagle> does that mean it will be
> supported with Angular 2.4.x?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/13294#issuecomment-277396651>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC5I5qobG-miUeBHcDb39Au7M6XiuYFks5rY8CLgaJpZM4LHQfe>
> .
>

mattlewis92 commented:
I've found that using typescript 2.1 + ngc to emit .metadata.json files for 3rd party libraries is the only place where typescript 2.1 causes issues. When trying to consume a library compiled with TS 2.1, the DI metadata doesn't seem to work properly and angular throws an exception (I can't remember the exact error message off the top of my head). This is really hard to debug where this problem comes from, as it's only noticeable when trying to consume the library from another app. I wonder if it's worth putting in a console warning if TS >= 2.1 is used and `skipMetadataEmit !== true` when using ngc?

mmc41 commented:
It is really difficult to find out the extend of actual/planned TS2.1+ support in angular (including AOT, CLI-tools etc). It would be nice if the angular 2 team would communicate something, since single issues like this do not paint the whole picture.

alexeagle commented:
I'll check with devrel about publishing something, but the simple answer
is, when ng4 comes out all the tooling will support the latest TS version,
and we may backport things to ng2 but we're not officially supporting it.

On Sun, Feb 5, 2017 at 12:46 PM Morten Christensen <notifications@github.com>
wrote:

> It is really difficult to find out the extend of actual/planned TS2.1+
> support in angular (including AOT, CLI-tools etc). It would be nice if the
> angular 2 team would communicate something, since single issues like this
> do not paint the whole picture.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/13294#issuecomment-277547843>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC5I1nBo_nO7Apv2vwy84cu9T5zWDMCks5rZjUIgaJpZM4LHQfe>
> .
>

mlc-mlapis commented:
Dears, can anybody do something with "good news and bad news"?

IgorMinar commented on build.sh:
```diff
@@ -161,18 +161,6 @@ do
     ${TSC} -p ${SRCDIR}/tsconfig-2015.json
   fi
 
-  echo "======      TSC 1.8 d.ts compat for ${DESTDIR}   ====="
```
removing this is a breaking change. can you please update the commit message to contain the "BREAKING CHANGE: " section per https://github.com/angular/angular/blob/master/CONTRIBUTING.md#footer

alexeagle replied:
done

IgorMinar commented on modules/@angular/compiler/src/expression_parser/lexer.ts:
```diff
@@ -294,6 +294,8 @@ class _Scanner {
         this.advance();
         let unescapedCode: number;
+        // Workaround for TS2.1-introduced type strictness
+        this.peek = this.peek;
```
does this do some kind of type casting or what?

IgorMinar replied:
Misko just hit this in g3 and we cherry-picked this hunk into his error unclassing PR

IgorMinar commented on tools/@angular/tsc-wrapped/test/typescript.mocks.ts:
```diff
@@ -119,6 +120,7 @@ export class MockSymbol implements ts.Symbol {
   getName(): string { return this.name; }
   getDeclarations(): ts.Declaration[] { return [this.node]; }
   getDocumentationComment(): ts.SymbolDisplayPart[] { return []; }
+  getJsDocTags(): ts.JSDocTagInfo[]{return []};
```
is this clang-formatted? it seems to be to crammed.

alexeagle replied:
maybe not

IgorMinar commented on tools/@angular/tsc-wrapped/test/tsc.spec.ts:
```diff
@@ -41,7 +41,7 @@ describe('options parsing', () => {
 
   it('should combine all options into ngOptions from vinyl like object', () => {
-    const {parsed, ngOptions} = pureTsc.readConfiguration(config as VinylFile, 'basePath');
+    const {parsed, ngOptions} = tsc.readConfiguration(config as VinylFile, 'basePath');
```
where does `tsc` come from? is it a global?

alexeagle replied:
it's imported above

alexeagle replied:
err, we create the tsc variable from a `new Tsc`

IgorMinar commented on tools/public_api_guard/core/index.d.ts:
```diff
@@ -334,64 +328,59 @@ export declare const CUSTOM_ELEMENTS_SCHEMA: SchemaMetadata;
     queryAll(predicate: Predicate<DebugElement>): DebugElement[];
     queryAllNodes(predicate: Predicate<DebugNode>): DebugNode[];
-    removeChild(child: DebugNode): void;
+    readonly children: DebugElement[];
```
this sorting is a bit unfortunate. we should fix this in ts-api-guardian after this PR lands

IgorMinar commented on tools/public_api_guard/core/index.d.ts:
```diff
@@ -419,9 +408,9 @@ export declare class ElementRef {
-    context: C;
-    rootNodes: any[];
+export declare abstract class EmbeddedViewRef<C> extends ViewRef { context: C;
+    readonly abstract
```
this looks wrong

IgorMinar commented on tools/public_api_guard/core/index.d.ts:
```diff
@@ -660,10 +649,10 @@ export declare abstract class NgModuleFactoryLoader {
+export declare abstract class NgModuleRef<T> { componentFactoryResolver: ComponentFactoryResolver; injector: Injector; instance: T;
+    readonly abstract
+    readonly abstract
+    readonly abstract
```
bug in ts-api-guardian?

IgorMinar commented on tools/public_api_guard/forms/index.d.ts:
```diff
@@ -393,52 +325,36 @@ export declare class NgControlStatusGroup extends AbstractControlStatus {
-    getControl(dir: NgModel): FormControl;
-    getFormGroup(dir: NgModelGroup): FormGroup;
-    onReset(): void;
-    onSubmit($event: Event): boolean;
```
what happened to all of these?

IgorMinar reviewed (changes requested):
I'm primarily concerned about the issues found in the public api guard files. Was that change generated after chuck upgraded our version of ts-api-guardian to be 2.1 compatible?

alexeagle commented:
updated to subsume #14287 for ts-api-guardian

alexeagle committed:
build: update public API with readonly
